compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java

```diff
@@ -3482,21 +3532,49 @@ public Boolean visit(BParameterizedType t, BType s) {
         }
     };
 
-    private boolean checkUnionHasSameType(LinkedHashSet<BType> memberTypes, BType baseType) {
+    private boolean isNil(BType type) {
+        BType referredType = getReferredType(type);
+        if (referredType.getKind() == TypeKind.NIL) {
+            return true;
+        } else if (referredType.getKind() == TypeKind.FINITE) {
+            Set<BLangExpression>  valueSpace = ((BFiniteType) referredType).getValueSpace();
+            return valueSpace.size() == 1 && valueSpace.iterator().next().getBType().getKind() == TypeKind.NIL;
+        }
+        return false;
+    }
+
+    private boolean checkUnionHasSameType(BUnionType unionType, BType baseType) {
+        LinkedHashSet<BType> memberTypes = unionType.getMemberTypes();
         boolean isSameType = false;
         for (BType type : memberTypes) {
             type = getReferredType(type);
             if (type.tag == TypeTags.FINITE) {
                 for (BLangExpression expr : ((BFiniteType) type).getValueSpace()) {
-                    isSameType = isSameOrderedType(expr.getBType(), baseType);
+                    isSameType = isSameOrderedType(expr.getBType(), baseType) || isNil(type);
                     if (!isSameType) {
                         return false;
                     }
                 }
             } else if (type.tag == TypeTags.UNION) {
-                return checkUnionHasSameType((LinkedHashSet<BType>) ((UnionType) type).getMemberTypes(), baseType);
+                if (checkUnionHasSameType((BUnionType) type, baseType)) {
+                    isSameType = true;
+                    continue;
+                }
+                return false;
```

Comment:
It will be more clear if we can do something like this,
```
if (!checkUnionHasSameType((BUnionType) type, baseType)) {
     return false;
}
isSameType = true;
```